core/src/main/java/com/datastax/oss/driver/internal/core/data/DefaultTupleValue.java

```diff
@@ -93,6 +98,78 @@ private void readObject(ObjectInputStream stream) throws InvalidObjectException
     throw new InvalidObjectException("Proxy required");
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (!(o instanceof TupleValue)) {
+      return false;
+    }
+    DefaultTupleValue that = (DefaultTupleValue) o;
+
+    if (!this.protocolVersion().equals(that.protocolVersion())) {
+      return false;
+    }
+
+    for (int i = 0; i < values.length; i++) {
+      DataType innerThisType = type.getComponentTypes().get(i);
+      DataType innerThatType = that.type.getComponentTypes().get(i);
+      if (!innerThisType.equals(innerThatType)) {
+        return false;
+      }
+      TypeCodec thatCodec = this.codecRegistry().codecFor(innerThatType);
+      TypeCodec thisCodec = that.codecRegistry().codecFor(innerThatType);
+      // If we can't find a needed codec, fallback on byte comparison
+      if (thatCodec == null || thisCodec == null) {
+        LOG.warn(
+            "No codec found for tuple type "
+                + innerThisType.toString()
+                + " falling back to byte comparison");
+        return Arrays.equals(values, that.values);
+      }
+      Object thisValue =
+          this.codecRegistry()
+              .codecFor(innerThisType)
+              .decode(this.values[i], this.protocolVersion());
+      Object thatValue =
+          that.codecRegistry()
+              .codecFor(innerThatType)
+              .decode(that.values[i], that.protocolVersion());
+      if (!Objects.equals(thisValue, thatValue)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public int hashCode() {
+
+    int result = Objects.hash(type);
```

Comment:
Since only one object is getting hashed here, this can be replaced by `type.hashCode()`.
Same at line 161.